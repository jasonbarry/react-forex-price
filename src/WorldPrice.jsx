// @flow
import * as React from 'react';
import * as Util from './utils';
import CURRENCY from './json/currencies.json';

type Props = {
  amount: number,
  baseCurrency: string,
  displayCurrency: string,
  hideCents: boolean,
  rounding: number => number,
  unwrap: boolean,
};

type State = {
  rates: Object,
};

export default class Price extends React.Component<Props, State> {
  static defaultProps = {
    baseCurrency: 'USD',
    displayCurrency: Util.getCurrencyFromBrowserLocale(),
    hideCents: false,
    rounding: Math.round,
    unwrap: false,
  };

  state = {
    rates: {},
  };

  componentDidMount() {
    this.fetchRates(this.props.baseCurrency);
  }

  // in case baseCurrency changes
  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.baseCurrency !== this.props.baseCurrency) {
      this.fetchRates(nextProps.baseCurrency);
    }
  }

  get amount(): number {
    const { amount, displayCurrency, hideCents, rounding } = this.props;
    const converted = parseFloat(amount) * (this.state.rates[displayCurrency] || 1);
    return hideCents ? rounding(converted) : converted;
  }

  fetchRates = (currency: string) => {
    // must check global variable so that multiple instances on same page
    // don't trigger their own fetch the first time the user loads the page that day
    if (window.__REACT_WORLD_PRICE_FETCHING__) {
      setTimeout(() => {
        this.fetchRates(currency);
      }, 50);
    } else {
      Util.fetchRates(currency).then(rates => { this.setState({ rates }); });
    }
  }

  render() {
    const { baseCurrency, displayCurrency, rounding, unwrap } = this.props;
    const amount = parseFloat(this.props.amount);

    // check for validity
    const invalidAmount = amount.toString() === 'NaN';
    const invalidBaseCurrency = !(baseCurrency in CURRENCY);
    const invalidDisplayCurrency = !(displayCurrency in CURRENCY);
    if (invalidAmount || invalidBaseCurrency || invalidDisplayCurrency) {
      console.error('react-world-price', this.props);
      return unwrap ? this.props.amount : <span>{this.props.amount}</span>;
    }

    const original = `${Util.formatAmountForCurrency(amount, CURRENCY[baseCurrency], rounding)} ${baseCurrency}`;
    const converted = Util.formatAmountForCurrency(this.amount, CURRENCY[displayCurrency], rounding);

    if (unwrap) return converted;
    return <span title={original}>{converted}</span>;
  }
}
