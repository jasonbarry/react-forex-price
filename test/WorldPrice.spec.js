import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import './setup';

import WorldPrice from '../src/WorldPrice';

describe('<WorldPrice />', () => {
  describe('Default props', () => {
    it('only amount passed in', () => {
      const instance = <WorldPrice amount={123} />;
      const wrapper = shallow(instance);
      const rendered = renderer.create(instance);
      expect(wrapper.text()).toEqual('$123');
      expect(rendered.toJSON()).toMatchSnapshot();
    });
    it('amount and baseCurrency', () => {
      const instance = <WorldPrice amount={123} baseCurrency="USD" />;
      const wrapper = shallow(instance);
      const rendered = renderer.create(instance);
      expect(wrapper.text()).toEqual('$123');
      expect(rendered.toJSON()).toMatchSnapshot();
    });
    it('amount and displayCurrency', () => {
      const instance = <WorldPrice amount={123} displayCurrency="EUR" />;
      const wrapper = shallow(instance);
      const rendered = renderer.create(instance);
      expect(wrapper.text()).toEqual('â‚¬123');
      expect(rendered.toJSON()).toMatchSnapshot();
    });
  });
  describe('Invalid inputs', () => {
    it('null amount returns empty string', () => {
      const instance = <WorldPrice amount={null} />;
      const wrapper = shallow(instance);
      const rendered = renderer.create(instance);
      expect(wrapper.text()).toEqual('');
      expect(rendered.toJSON()).toMatchSnapshot();
    });
    it('undefined amount returns empty string', () => {
      const instance = <WorldPrice amount={undefined} />;
      const wrapper = shallow(instance);
      const rendered = renderer.create(instance);
      expect(wrapper.text()).toEqual('');
      expect(rendered.toJSON()).toMatchSnapshot();
    });
    it('string returns original string', () => {
      const instance = <WorldPrice amount={'hello'} />;
      const wrapper = shallow(instance);
      const rendered = renderer.create(instance);
      expect(wrapper.text()).toEqual('hello');
      expect(rendered.toJSON()).toMatchSnapshot();
    });
    it('unsupported base currency returns original amount', () => {
      const instance = <WorldPrice amount={987.234} baseCurrency="" />;
      const wrapper = shallow(instance);
      const rendered = renderer.create(instance);
      expect(wrapper.text()).toEqual('987.234');
      expect(rendered.toJSON()).toMatchSnapshot();
    });
    it('unsupported display currency returns original amount', () => {
      const instance = <WorldPrice amount={987.234} displayCurrency="" />;
      const wrapper = shallow(instance);
      const rendered = renderer.create(instance);
      expect(wrapper.text()).toEqual('987.234');
      expect(rendered.toJSON()).toMatchSnapshot();
    });
  });
  describe('Rounding', () => {
    it('no rounding prop passed in', () => {
      const wrapper = shallow(<WorldPrice amount={8675.309} />);
      expect(wrapper.text()).toEqual('$8,675.31');
    });
    it('rounding up', () => {
      const wrapper = shallow(<WorldPrice amount={8675.309} rounding={Math.ceil} />);
      expect(wrapper.text()).toEqual('$8,675.31');
    });
    it('rounding down', () => {
      const wrapper = shallow(<WorldPrice amount={8675.309} rounding={Math.floor} />);
      expect(wrapper.text()).toEqual('$8,675.30');
    });
    it('rounding up with hideCents', () => {
      const wrapper = shallow(<WorldPrice amount={8675.309} rounding={Math.ceil} hideCents />);
      expect(wrapper.text()).toEqual('$8,676');
    });
    it('rounding down with hideCents', () => {
      const wrapper = shallow(<WorldPrice amount={8675.309} rounding={Math.floor} hideCents />);
      expect(wrapper.text()).toEqual('$8,675');
    });
  });
});
