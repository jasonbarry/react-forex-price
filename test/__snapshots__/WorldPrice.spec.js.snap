// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WorldPrice /> Default props amount and baseCurrency 1`] = `
<span
  title="$123 USD"
>
  $123
</span>
`;

exports[`<WorldPrice /> Default props amount and displayCurrency 1`] = `
<span
  title="$123 USD"
>
  â‚¬123
</span>
`;

exports[`<WorldPrice /> Default props only amount passed in 1`] = `
<span
  title="$123 USD"
>
  $123
</span>
`;

exports[`<WorldPrice /> Invalid inputs null amount returns empty string 1`] = `<span />`;

exports[`<WorldPrice /> Invalid inputs string returns original string 1`] = `
<span>
  hello
</span>
`;

exports[`<WorldPrice /> Invalid inputs undefined amount returns empty string 1`] = `<span />`;

exports[`<WorldPrice /> Invalid inputs unsupported base currency returns original amount 1`] = `
<span>
  987.234
</span>
`;

exports[`<WorldPrice /> Invalid inputs unsupported display currency returns original amount 1`] = `
<span>
  987.234
</span>
`;
